docs/installation-configuration.md
---
title: Datadog Provider Installation & Configuration
meta_desc: Provides an overview on how to configure the Pulumi Datadog Provider.
layout: package
---
## Installation

The datadog provider is available as a package in all Pulumi languages:

* JavaScript/TypeScript: [`@pulumi/datadog`](https://www.npmjs.com/package/@pulumi/datadog)
* Python: [`pulumi-datadog`](https://pypi.org/project/pulumi-datadog/)
* Go: [`github.com/pulumi/pulumi-datadog/sdk/v4/go/datadog`](https://github.com/pulumi/pulumi-datadog)
* .NET: [`Pulumi.Datadog`](https://www.nuget.org/packages/Pulumi.Datadog)
* Java: [`com.pulumi/datadog`](https://central.sonatype.com/artifact/com.pulumi/datadog)

The [Datadog](https://www.datadoghq.com) provider is used to interact with the resources supported by Datadog. The provider needs to be configured with the proper credentials before it can be used.

Use the navigation to the left to read about the available resources.
## Example Usage

{{< chooser language "typescript,python,go,csharp,java,yaml" >}}
{{% choosable language typescript %}}
```yaml
# Pulumi.yaml provider configuration file
name: configuration-example
runtime: nodejs
config:
    datadog:apiKey:
        value: 'TODO: var.datadog_api_key'
    datadog:appKey:
        value: 'TODO: var.datadog_app_key'

```

{{% /choosable %}}
{{% choosable language python %}}
```yaml
# Pulumi.yaml provider configuration file
name: configuration-example
runtime: python
config:
    datadog:apiKey:
        value: 'TODO: var.datadog_api_key'
    datadog:appKey:
        value: 'TODO: var.datadog_app_key'

```

{{% /choosable %}}
{{% choosable language csharp %}}
```yaml
# Pulumi.yaml provider configuration file
name: configuration-example
runtime: dotnet
config:
    datadog:apiKey:
        value: 'TODO: var.datadog_api_key'
    datadog:appKey:
        value: 'TODO: var.datadog_app_key'

```

{{% /choosable %}}
{{% choosable language go %}}
```yaml
# Pulumi.yaml provider configuration file
name: configuration-example
runtime: go
config:
    datadog:apiKey:
        value: 'TODO: var.datadog_api_key'
    datadog:appKey:
        value: 'TODO: var.datadog_app_key'

```

{{% /choosable %}}
{{% choosable language yaml %}}
```yaml
# Pulumi.yaml provider configuration file
name: configuration-example
runtime: yaml
config:
    datadog:apiKey:
        value: 'TODO: var.datadog_api_key'
    datadog:appKey:
        value: 'TODO: var.datadog_app_key'

```

{{% /choosable %}}
{{% choosable language java %}}
```yaml
# Pulumi.yaml provider configuration file
name: configuration-example
runtime: java
config:
    datadog:apiKey:
        value: 'TODO: var.datadog_api_key'
    datadog:appKey:
        value: 'TODO: var.datadog_app_key'

```

{{% /choosable %}}
{{< /chooser >}}

<!-- schema generated by tfplugindocs -->
## Schema
### Optional

- `apiKey` (String, Sensitive) (Required unless validate is false) Datadog API key. This can also be set via the DD_API_KEY environment variable.
- `apiUrl` (String) The API URL. This can also be set via the DD_HOST environment variable, and defaults to `https://api.datadoghq.com`. Note that this URL must not end with the `/api/` path. For example, `https://api.datadoghq.com/` is a correct value, while `https://api.datadoghq.com/api/` is not. And if you're working with "EU" version of Datadog, use `https://api.datadoghq.eu/`. Other Datadog region examples: `https://api.us5.datadoghq.com/`, `https://api.us3.datadoghq.com/` and `https://api.ddog-gov.com/`. See <https://docs.datadoghq.com/getting_started/site/> for all available regions.
- `appKey` (String, Sensitive) (Required unless validate is false) Datadog APP key. This can also be set via the DD_APP_KEY environment variable.
- `defaultTags` (Block List, Max: 1) [Experimental - Monitors only] Configuration block containing settings to apply default resource tags across all resources. (see below for nested schema)
- `httpClientRetryBackoffBase` (Number) The HTTP request retry back off base. Defaults to 2.
- `httpClientRetryBackoffMultiplier` (Number) The HTTP request retry back off multiplier. Defaults to 2.
- `httpClientRetryEnabled` (String) Enables request retries on HTTP status codes 429 and 5xx. Valid values are [`true`, `false`]. Defaults to `true`.
- `httpClientRetryMaxRetries` (Number) The HTTP request maximum retry number. Defaults to 3.
- `httpClientRetryTimeout` (Number) The HTTP request retry timeout period. Defaults to 60 seconds.
- `validate` (String) Enables validation of the provided API key during provider initialization. Valid values are [`true`, `false`]. Default is true. When false, apiKey won't be checked.

<a id="nestedblock--default_tags"></a>
### Nested Schema for `defaultTags`

Optional:

- `tags` (Map of String) [Experimental - Monitors only] Resource tags to be applied by default across all resources.

CHANGELOG.md
CHANGELOG
=========

## Notice (2022-01-06)

*As of this notice, using CHANGELOG.md is DEPRECATED. We will be using [GitHub Releases](https://github.com/pulumi/pulumi-datadog/releases) for this repository*

## HEAD (Unreleased)
_(none)_

---

## 4.5.0 (2021-12-20)
* Upgrade to v3.7.0 of the DataDog Terraform Provider
* Upgrade to v3.13.0 of the Pulumi Terraform Bridge
* Upgrade to v3.19.0 of the Pulumi SDK

## 4.4.0 (2021-11-18)
* Upgrade to terraform-bridge 3.11.0
* Upgrade to pulumi 3.17.0

## 4.3.0 (2021-10-29)
* Upgrade to v3.5.0 of the datadog Terraform Provider

## 4.2.0 (2021-09-27)
* Upgrade to v3.4.0 of the datadog Terraform Provider

## 4.1.0 (2021-09-03)
* Upgrade to v3.3.0 of the datadog Terraform Provider

## 4.0.0 (2021-07-12)
* Upgrade to v3.1.2 of the datadog Terraform Provider  
  **PLEASE NOTE the following breaking changes:**  
  * `datadog.TimeBoard` and `datadog.Screenboard` resources have been removed as they were previously deprecated.
  * `datadog.Monitor` has had `threshold`, `thresholdWindows` and `silenced` parameters removed
  * `datadog.User` has had `handle`, `isAdmin`, `accessRole` and `role` parameters removed
  * `datadog.syntheticsTest` has had `request`, `assertions`, `options`, `step` and `variable` parameters removed
  * `datadog.logsArchive` has had `s3`, `azure`, and `gcp` parameters removed
  * `datadog.serviceLevelObjective` has had `monitorSearch` parameter removed
  * `datadog.PagerDuty.Integration` has had `services` parameter removed
  * `datadog.dashboard` has had `layout`, `time`, `logset`, `count`, `start`, `compute` and `search` parameters removed

## 3.3.0 (2021-05-27)
* Upgrade to v3.2.1 of the pulumi-terraform-bridge

## 3.2.1 (2021-05-27)
* Upgrade to v2.26.1 of the datadog Terraform Provider

## 3.2.0 (2021-05-21)
* Upgrade to v2.26.0 of the datadog Terraform Provider

## 3.1.0 (2021-05-12)
* Upgrade to v2.25.0 of the datadog Terraform Provider

## 3.0.0 (2021-04-19)
* Depend on Pulumi 3.0, which includes improvements to Python resource arguments and key translation, Go SDK performance,
  Node SDK performance, general availability of Automation API, and more.

## 2.18.0 (2021-04-12)
* Upgrade to pulumi-terraform-bridge v2.23.0

## 2.17.1 (2021-03-23)
* Upgrade to pulumi-terraform-bridge v2.22.1  
  **Please Note:** This includes a bug fix to the refresh operation regarding arrays

## 2.17.0 (2021-03-15)
* Upgrade to pulumi-terraform-bridge v2.21.0
* Release macOS arm64 binary

## 2.16.1 (2021-02-16)
* Upgrade to pulumi-terraform-bridge v2.19.0  
  **Please Note:** This includes a bug fix that stops mutating resources options in the nodejs provider
* Avoid storing config from the environment into the state

## 2.16.0 (2021-01-29)
* Upgrade to pulumi-terraform-bridge v2.18.1

## 2.15.0 (2021-01-21)
* Upgrade to v2.20.0 of the datadog Terraform Provider

## 2.14.1 (2021-01-13)
* Upgrade to pulumi-terraform-bridge v2.17.0
* Upgrade to Pulumi v2.17.0

## 2.14.0 (2021-01-08)
* Upgrade to v2.19.1 of the datadog Terraform Provider

## 2.13.0 (2020-12-10)
* Upgrade to v2.18.1.0 of the datadog Terraform Provider
* Upgrade to pulumi-terraform-bridge v2.15.3

## 2.12.0 (2020-12-08)
* Upgrade to v2.17.0 of the datadog Terraform Provider

## 2.11.0 (2020-11-24)
* Upgrade to v2.16.0 of the datadog Terraform Provider
* Upgrade to pulumi-terraform-bridge v2.13.2  
  * This adds support for import specific examples in documentation

## 2.10.0 (2020-10-28)
* Upgrade to v2.14.0 of the datadog Terraform Provider

## 2.9.0 (2020-10-26)
* Upgrade to Pulumi v2.12.0 and pulumi-terraform-bridge v2.11.0
* Improving the accuracy of previews leading to a more accurate understanding of what will actually change rather than assuming all output properties will change.  
  ** PLEASE NOTE:**  
  This new preview functionality can be disabled by setting `PULUMI_DISABLE_PROVIDER_PREVIEW` to `1` or `false`.

## 2.8.0 (2020-09-21)
* Upgrade to v2.13.0 of the datadog Terraform Provider
* Upgrade to Pulumi v2.10.0
* Upgrade to pulumi-terraform-bridge v2.8.0

## 2.7.0 (2020-08-31)
* Upgrade to pulumi-terraform-bridge v2.7.3
* Upgrade to Pulumi v2.9.0, which adds type annotations and input/output classes to Python

## 2.6.0 (2020-07-23)
* Upgrade to v2.12.0 of the datadog Terraform Provider

## 2.5.0 (2020-06-30)
* Upgrade to v2.11.0 of the datadog Terraform Provider

## 2.4.0 (2020-06-26)
* Upgrade to v2.10.0 of the datadog Terraform Provider

## 2.3.0 (2020-06-23)
* Upgrade to v2.9.0 of the datadog Terraform Provider

## 2.2.0 (2020-06-16)
* Upgrade to v2.8.0 of the datadog Terraform provider.
* Switch to GitHub actions for build

## 2.1.2 (2020-05-28)
* Upgrade to Pulumi v2.3.0
* Upgrade to pulumi-terraform-bridge v2.4.0

## 2.1.1 (2020-05-12)
* Upgrade to pulumi-terraform-bridge v2.3.1

## 2.1.0 (2020-04-28)
* Regenerate datasource examples to be async
* Upgrade to pulumi-terraform-bridge v2.1.0

## 2.0.0 (2020-04-17)
* Upgrade to Pulumi v2.0.0
* Upgrade to pulumi-terraform-bridge v2.0.0

## 1.4.0 (2020-04-14)
* Upgrade to Pulumi v1.13.1
* Upgrade to pulumi-terraform-bridge v1.8.4
* Refactor layout to support Go modules

## 1.3.0 (2020-03-23)
* Namespace names in .NET SDK are adjusted to PascalCase
([#33](https://github.com/pulumi/pulumi-datadog/pull/33)).
* Upgrade to v2.7.0 of the datadog Terraform provider.
* Upgrade to Pulumi v1.12.1
* Upgrade to pulumi-terraform-bridge v1.8.2

## 1.2.0 (2019-12-04)
* Upgrade to support go 1.13.x
* Upgrade to pulumi-terraform@dotnet

## 1.1.0 (2019-11-15)
* Add support for DotNet SDK Generation

## 1.0.0 (2019-10-23)
* Regenerate SDK based on tf2pulumi 0.6.0
* Upgrade to v2.5.0 of the datadog Terraform provider.

## 0.18.11 (2019-09-11)
* Upgrade to v2.4.0 of the datadog Terraform provider.

## 0.18.10 (2019-09-05)
* Upgrade to Pulumi v1.0.0

## 0.18.9 (2019-08-30)
* Upgrade to v2.3.0 of the datadog Terraform provider.

## 0.18.8 (2019-08-29)
* Upgrade to v2.2.0 of the datadog Terraform provider.

sdk/dotnet/Config/Config.cs
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***

using System;
using System.Collections.Immutable;

namespace Pulumi.Datadog
{
    public static class Config
    {
        [global::System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "IDE1006", Justification = 
        "Double underscore prefix used to avoid conflicts with variable names.")]
        private sealed class __Value<T>
        {
            private readonly Func<T> _getter;
            private T _value = default!;
            private bool _set;

            public __Value(Func<T> getter)
            {
                _getter = getter;
            }

            public T Get() => _set ? _value : _getter();

            public void Set(T value)
            {
                _value = value;
                _set = true;
            }
        }

        private static readonly global::Pulumi.Config __config = new global::Pulumi.Config("datadog");

        private static readonly __Value<string?> _apiKey = new __Value<string?>(() => __config.Get("apiKey"));
        /// <summary>
        /// (Required unless validate is false) Datadog API key. This can also be set via the DD_API_KEY environment variable.
        /// </summary>
        public static string? ApiKey
        {
            get => _apiKey.Get();
            set => _apiKey.Set(value);
        }

        private static readonly __Value<string?> _apiUrl = new __Value<string?>(() => __config.Get("apiUrl"));
        /// <summary>
        /// The API URL. This can also be set via the DD_HOST environment variable, and defaults to `https://api.datadoghq.com`.
        /// Note that this URL must not end with the `/api/` path. For example, `https://api.datadoghq.com/` is a correct value,
        /// while `https://api.datadoghq.com/api/` is not. And if you're working with "EU" version of Datadog, use
        /// `https://api.datadoghq.eu/`. Other Datadog region examples: `https://api.us5.datadoghq.com/`,
        /// `https://api.us3.datadoghq.com/` and `https://api.ddog-gov.com/`. See https://docs.datadoghq.com/getting_started/site/
        /// for all available regions.
        /// </summary>
        public static string? ApiUrl
        {
            get => _apiUrl.Get();
            set => _apiUrl.Set(value);
        }

        private static readonly __Value<string?> _appKey = new __Value<string?>(() => __config.Get("appKey"));
        /// <summary>
        /// (Required unless validate is false) Datadog APP key. This can also be set via the DD_APP_KEY environment variable.
        /// </summary>
        public static string? AppKey
        {
            get => _appKey.Get();
            set => _appKey.Set(value);
        }

        private static readonly __Value<Pulumi.Datadog.Config.Types.DefaultTags?> _defaultTags = new __Value<Pulumi.Datadog.Config.Types.DefaultTags?>(() => __config.GetObject<Pulumi.Datadog.Config.Types.DefaultTags>("defaultTags"));
        /// <summary>
        /// [Experimental - Monitors only] Configuration block containing settings to apply default resource tags across all
        /// resources.
        /// </summary>
        public static Pulumi.Datadog.Config.Types.DefaultTags? DefaultTags
        {
            get => _defaultTags.Get();
            set => _defaultTags.Set(value);
        }

        private static readonly __Value<int?> _httpClientRetryBackoffBase = new __Value<int?>(() => __config.GetInt32("httpClientRetryBackoffBase"));
        /// <summary>
        /// The HTTP request retry back off base. Defaults to 2.
        /// </summary>
        public static int? HttpClientRetryBackoffBase
        {
            get => _httpClientRetryBackoffBase.Get();
            set => _httpClientRetryBackoffBase.Set(value);
        }

        private static readonly __Value<int?> _httpClientRetryBackoffMultiplier = new __Value<int?>(() => __config.GetInt32("httpClientRetryBackoffMultiplier"));
        /// <summary>
        /// The HTTP request retry back off multiplier. Defaults to 2.
        /// </summary>
        public static int? HttpClientRetryBackoffMultiplier
        {
            get => _httpClientRetryBackoffMultiplier.Get();
            set => _httpClientRetryBackoffMultiplier.Set(value);
        }

        private static readonly __Value<string?> _httpClientRetryEnabled = new __Value<string?>(() => __config.Get("httpClientRetryEnabled"));
        /// <summary>
        /// Enables request retries on HTTP status codes 429 and 5xx. Valid values are [`true`, `false`]. Defaults to `true`.
        /// </summary>
        public static string? HttpClientRetryEnabled
        {
            get => _httpClientRetryEnabled.Get();
            set => _httpClientRetryEnabled.Set(value);
        }

        private static readonly __Value<int?> _httpClientRetryMaxRetries = new __Value<int?>(() => __config.GetInt32("httpClientRetryMaxRetries"));
        /// <summary>
        /// The HTTP request maximum retry number. Defaults to 3.
        /// </summary>
        public static int? HttpClientRetryMaxRetries
        {
            get => _httpClientRetryMaxRetries.Get();
            set => _httpClientRetryMaxRetries.Set(value);
        }

        private static readonly __Value<int?> _httpClientRetryTimeout = new __Value<int?>(() => __config.GetInt32("httpClientRetryTimeout"));
        /// <summary>
        /// The HTTP request retry timeout period. Defaults to 60 seconds.
        /// </summary>
        public static int? HttpClientRetryTimeout
        {
            get => _httpClientRetryTimeout.Get();
            set => _httpClientRetryTimeout.Set(value);
        }

        private static readonly __Value<string?> _validate = new __Value<string?>(() => __config.Get("validate"));
        /// <summary>
        /// Enables validation of the provided API key during provider initialization. Valid values are [`true`, `false`]. Default
        /// is true. When false, api_key won't be checked.
        /// </summary>
        public static string? Validate
        {
            get => _validate.Get();
            set => _validate.Set(value);
        }

        public static class Types
        {

             public class DefaultTags
             {
            /// <summary>
            /// [Experimental - Monitors only] Resource tags to be applied by default across all resources.
            /// </summary>
                public ImmutableDictionary<string, string>? Tags { get; set; } = null!;
            }
        }
    }
}
